Add setting bios.shell_path

hugeblank opened the pull request:
Allows the user to directly set the path the bios uses to launch the shell. Setting `bios.shell_path` takes precedent over `bios.use_multishell`. Essentially a custom shell may choose to implement `shell` and `multishell` APIs, independent of whether use_multishell is set. This is loosely reflected in the description of the setting. Also some additional status checking has been implemented to allow for shell errors to be easily read before computer shutdown. 

My justification for adding this is that while [unbios](https://gist.github.com/MCJack123/42bc69d3757226c966da752df80437dc) exists, it is the logical extreme of what I imagine most players want. This PR is the spirit of unbios in a way that makes sense for CC, that advanced players can take advantage of (if I'm not mistaken mbs could benefit from this too, no?) Also servers presumably would ever so slightly benefit from computers that don't *need* a whole shell having the ability to start without one.

hugeblank committed:
Add setting bios.shell_path

Allow the user to directly set the path the bios uses for the shell.

Has some additional status checking to allow for shell errors to be read before computer shutdown.

hugeblank committed:
trailing comma

I learned how to lintLua!

Wojbie commented on projects/core/src/main/resources/data/computercraft/lua/bios.lua:
```diff
@@ -728,24 +742,24 @@ if fs.exists(".settings") then
-        else
-            sShell = "rom/programs/shell.lua"
+        shellOk = os.run({}, settings.get("bios.shell_path"))
+        if shellOk then
```
Is there a reason this part of logic is changed and extra variable is checked?

hugeblank replied:
yes, when the shell errors os.run emits false, and does not error. this was never caught before and so after 1 second of the error displaying, the computer shuts down. The only reason we see the error in the first place is because we're afforded that 1 second by the shutdown rom program saying goodbye. My change allows a user to see the error at their leisure, where the "press any key" is.

hugeblank replied:
This actually runs a little deeper, in that we don't know on what terminal the error from os.run was rendered on. Ideally we'd redirect to the native one, like we do when the whole parallel loop errors. I almost wish os.run gave the error string like pcall does so we could just `printError` it like we do with the parallel loop.

edit - os.run is defined in bios, lucky me!

hugeblank replied:
latest commit extracts os.run logic so we're able to print the error on the correct terminal

SquidDev commented:
This was previously discussed in #710 and #911, and it's still not clear to me if this is actually useful. I understand the instinctual desire for this, but I don't actually think it ends up solving any practical problem.

hugeblank commented:
The utility for me is that I don't have to deal with my own shell (or programs that don't require a shell) being layered atop an existing shell. It just feels like an extra step that should have the ability to be removed, especially when it's dead simple to implement, does not break compatibility, and is entirely up to the user to take advantage of.

Wojbie commented:
How does the dead shell impacts anything? Assuming multishell is off and you called `shell.exit()` in startup code that then initiated your shell it no longer does anything and will exit once your code exits.

hugeblank commented:
> Assuming multishell is off and you called `shell.exit()` in startup code that then initiated your shell it no longer does anything and will exit once your code exits.

Startup programs wouldn't even be ran, and the shell API would not exist. Those are both things the shell does. Basically any `shell.<...>` setting would become something the user may (or may not) implement in their custom shell.

hugeblank committed:
Correctly display shell errors on native term

Extracts os.run into an internal launch function so we can extract shell errors and correctly display them on the native terminal.
